core/src/main/scala/org/apache/spark/executor/Executor.scala

```diff
@@ -218,6 +244,8 @@ private[spark] class Executor(
     env.metricsSystem.report()
     heartbeater.shutdown()
     heartbeater.awaitTermination(10, TimeUnit.SECONDS)
+    executorPluginThread.interrupt()
+    executorPluginThread.join()
```

Comment:
We should be careful about semantics of `interrupt()` when overriding it - here we are using it to trigger something in addition to thread interruption.
For example, the actual shutdown of plugin's is getting triggered in this caller thread - and not in plugin init thread : which can cause issues since plugin's are using a a different thread ctx classloader.

This usage is actually a bit confusing.
A better option, if we want to follow current design, would be:
* Subclass Thread - expose methods to trigger state changes.
* Once init completes, wait+notify/condition wait until plugin shutdown is required.
* On stop(), invoke plugin shutdown in plugin thread (so that right classloader is in use).

IIRC VM shutdown results in `executor.stop` - and so plugin shutdown should get invoked even in that case.

Reply:
Oops, yeah, my bad. I believe the current implementation also terminates the plugins immediately...

For your recommendation, would the thread that initializes the plugins also be undergoing the conditional wait? If so, would that mean I cannot join on this thread right after initialization of plugins (from your earlier comment) since the thread would stay alive to wait for the shutdown condition?